cpp/ql/src/experimental/Security/CWE/CWE-14/CompilerRemovalOfCodeToClearBuffers.ql

```diff
@@ -0,0 +1,121 @@
+/**
+ * @name Compiler Removal Of Code To Clear Buffers
+ * @description --Using the memset function to clear private data as a final expression when working with a variable is potentially dangerous because the compiler can optimize this call.
+ *              --For some compilers, optimization is also possible when using calls to free memory after the memset function.
+ *              --To clear it, you need to use the RtlSecureZeroMemory or memset_s functions, or compilation flags that exclude optimization of memset calls (-fno-builtin-memset).
+ * @kind problem
+ * @id cpp/compiler-removal-of-code-to-clear-buffers
+ * @problem.severity warning
+ * @precision medium
+ * @tags security
+ *       external/cwe/cwe-14
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.DataFlow
+
+/**
+ * A call to `memset` , for some local variable.
+ */
+class CompilerRemovaMemset extends FunctionCall {
+  CompilerRemovaMemset() {
+    this.getTarget().hasName("memset") and
+    exists(DataFlow::Node source, DataFlow::Node sink, LocalVariable isv, Expr exp |
+      DataFlow::localFlow(source, sink) and
+      this.getArgument(0) = isv.getAnAccess() and
+      source.asExpr() = exp and
+      exp.getLocation().getEndLine() < this.getArgument(0).getLocation().getStartLine() and
+      sink.asExpr() = this.getArgument(0)
+    )
+  }
+
+  predicate isExistsAllocForThisVariable() {
+    exists(FunctionCall alloc, Variable v |
+      alloc = v.getAnAssignedValue() and
+      this.getArgument(0) = v.getAnAccess() and
+      alloc.getASuccessor+() = this
```

Comment:
Generally, doing control-flow analysis on individual expressions is very expensive. I think what you have here is fine, but there are a couple of improvements that you could make in the future:
- Use [BasicBlock](https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/controlflow/BasicBlocks.qll/type.BasicBlocks$BasicBlock.html)s instead. They also have a [getASuccessor](https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/controlflow/BasicBlocks.qll/predicate.BasicBlocks$BasicBlock$getASuccessor.0.html) predicate that is much faster.
- I get the impression that what you really want here is local dataflow from an allocation to the first argument of this `memset`. That will also be faster than doing expression-level control-flow analysis.

Reply:
Thank you.
However, my idea was to minimize the number of false detections as much as possible. It seemed to me that the flow of control would eliminate complex cases of access to a variable that occur outside the function. (for example, when memory was allocated outside the limits of the current function, and the managed variable received this pointer not from the function argument directly).

so if you agree, I would leave that as it is.

Reply:
I'm all for minimizing the number of false positives, so if you observe fewer false positives by going with control-flow instead of dataflow that's totally fine by me. 👍